### PR TITLE
Forward Message

### LINKED ISSUE #473: Add a forward message button
**Is your feature request related to a problem? Please describe.**

It is useful to be able to forward the contents of a message in one room to another room. I find this particularly useful for images, where you don't have to download and re-upload the image if it is sent into the new room with the same mxc url.

**Describe the solution you'd like**

Add a forward message button with roughly the same features as Element Web.

### PR DESCRIPTION
**Pull request by Jedi18**
Add functionality for forwarding a message (image,video,text..etc) from one room to another.
Adds a context menu item for forwarding a message.

Fixes #473

### CODE REVIEW

File: src/timeline/TimelineViewManager.cpp [hunk 1]
```diff
@@ -614,3 +618,98 @@ TimelineViewManager::focusTimeline()
...
+void
+TimelineViewManager::forwardMessageToRoom(mtx::events::collections::TimelineEvents *e,
+                                          QString roomId,
+                                          bool sentFromEncrypted)
+{
+        auto elem        = *e;
+        auto room        = models.find(roomId);
+        auto messageType = mtx::accessors::msg_type(elem);
+
+        if (sentFromEncrypted && messageType == mtx::events::MessageType::Image) {
```
Comment (deepbluev7) [inline src/timeline/TimelineViewManager.cpp:631]: This partially applies to all file types. And you don't need to recalculate anything, just upload the file object to generate a an url object instead, everything else can be just taken from the old event. Basically you can just copy the old info object, and replace file with nullopt and set the url instead to the newly uploaded file.

Comment (Jedi18) [reply src/timeline/TimelineViewManager.cpp]: Thanks! I've made the changes, should be ready for review now.

**Comment by deepbluev7** [comment]
Can you fix the compilation issues? :3

**Comment by Jedi18** [comment]
Oh right sorry, I thought the build-macos always failed. I dind't realize you'd fixed it :D

**Comment by deepbluev7** [comment]
Okay, finally got a chance to try it. It seems to work pretty well, but the design needs some work:

image

Do you want to do that or should I?

File: src/timeline/TimelineViewManager.h [hunk 2]
```diff
@@ -146,10 +147,58 @@ public slots:
...
+                }
+
+                if constexpr (is_detected<formatted_body_t, T>::value) {
+                        if (e.content.format == "org.matrix.custom.html") {
+                                e.content.formatted_body =
+                                  utils::stripReplyFromFormattedBody(e.content.formatted_body);
+                        }
+                }
+        }
+
```
Comment (deepbluev7) [inline src/timeline/TimelineViewManager.h:201]: This should probably go into mtx::accerrors or in the cpp file.

**Comment by Jedi18** [comment]
@deepbluev7 Thanks! Up to you, it might take me a while to do it since my exams start day after (though if I find time in between I may be able to do it). Let me know if you're doing it so we both don't end up working on it.

**Comment by deepbluev7** [comment]
Then let me fix it up, focus on your exams! :3

**Comment by Jedi18** [comment]
Ok, thanks a lot! :)

**Comment by deepbluev7** [comment]
Thank you for the contribution! <3

**Comment by Jedi18** [comment]
Thanks for fixing up the issues and the design :D

**Comment by deepbluev7** [comment]
Well, design still has some issues, but it is good enough for now xD

**Pull request merged**